static_show: Don't print field names of structs in `--trace-compile` output

topolarity opened the pull request:
This is a revival of https://github.com/JuliaLang/julia/pull/40652, to remove spurious reasons that `--trace-compile` output does not `eval` properly right now (I don't know if we'll be able to fix all of the flaws, but it seems worthwhile to fix the easy cases so that we can focus on the hard ones)

The default REPL printing does not think it's worth printing these field names:
```julia
julia> struct Foo
           x::Int
           y::Int
           z::Int
       end

julia> Foo(1,2,3)
Foo(1, 2, 3)
```
so I'm not sure why our `static_show` would find it necessary.

topolarity committed:
static_show: Don't print field names of structs

topolarity commented:
@vtjnash if you really want these field names and also don't want them printed as a comment (e.g. `Foo(#= x =# 1, #= y =# 2, #= z =# 3)`), I'll happily add this as a printing option in the config / context

vtjnash commented:
Debugging relies very heavily on these field names, so NAK any removal of them (it also makes the correct version of precompile worse, since it can no longer use those comments to try to construct the new object correctly)

topolarity commented:
Any complaint if they are printed as comments?

topolarity committed:
static_show: Add `suppress_field_names` config option

Also tweak the printing of field names to use comments:
"Foo(#= x =# 1, #= y =# 2, #= z =# 3)"

topolarity commented:
Added a config option. `jl_` now prints:
```julia
julia> ccall(:jl_, Cvoid, (Any,), Foo(1,2,3))
Main.Foo(#= x =# 1, #= y =# 2, #= z =# 3)
```

And `--trace-compile` prints:
```julia
precompile(Tuple{typeof(Main.foo), Base.Val{Main.Foo(1, 2, 3)}})
```

vtjnash commented:
Eh, it looks worse and makes implementing the good version much harder, but it is valid
